File: 资料/算法学习资料/数据结构专题/program/program4.pas
const 
	maxn = 5011;
	inf = $3F3F3F3F;
var
	n, m, data_type: longint;
	data: array [0..maxn] of array [0..maxn] of boolean;
	seed: longint;

function next_rand():longint;
const
	P = 1000000007;
	Q = 83978833;
	R = 8523467;
begin
	seed := (int64(Q) * seed mod P * seed + R) mod P;
	exit(seed);
end;

procedure generate_input();
var
	i, j: longint;
begin
	read(n, m, data_type);
	for i := 0 to n - 1 do
		for j := 0 to m - 1 do
			data[i][j] := ((next_rand() mod 8) > 0)
end;

function count1():int64;
var
	i, j, k, l: longint;
	ans: int64;
begin
	ans := 0;
	for i := 0 to n - 1 do
		for j := 0 to m - 1 do
			if data[i][j] then
				for k := 0 to n - 1 do
					for l := 0 to m - 1 do
						if data[k][l] and ((k <> i) or (l <> j)) then
							inc(ans);
	exit(ans)
end;

function abs_int(x: longint):longint;
begin
	if x < 0 then
		exit(-x)
	else
		exit(x)
end;

function count2():int64;
var
	i, j, k, l, level, dist: longint;
	ans: int64;
begin
	ans := 0;
	for i := 0 to n - 1 do
		for j := 0 to m - 1 do
			if data[i][j] then begin
				level := inf;
				for k := 0 to n - 1 do
					for l := 0 to m - 1 do
						if not data[k][l] then begin
							dist := abs_int(k - i) + abs_int(l - j);
							if level > dist then
								level := dist;
						end;
				inc(ans, level);
			end;
	exit(ans)
end;

procedure main();
var
	i: longint;
begin
	read(seed);
	for i := 0 to 9 do begin
		generate_input;
		if data_type = 0 then
			writeln(count1())
		else
			writeln(count2());
	end
end;

begin
	main;
end.

File: 资料/算法学习资料/数据结构专题/program/program4.c
#include <stdio.h>

#define N 5011
const int inf = 0x3F3F3F3F;
int n, m, type;
char data[N][N];

int seed;
int next_rand(){
	static const int P = 1000000007, Q = 83978833, R = 8523467;
	return seed = ((long long)Q * seed % P * seed + R) % P;
}

void generate_input(){
	int i, j;
	scanf("%d%d%d", &n, &m, &type);
	for(i = 0; i < n; i++)
		for(j = 0; j < m; j++)
			data[i][j] = ((next_rand() % 8) > 0 ? 1 : 0);
}

long long count1(){
	int i, j, k, l;
	long long ans = 0LL;
	for(i = 0; i < n; i++)
		for(j = 0; j < m; j++)
			if(data[i][j])
				for(k = 0; k < n; k++)
					for(l = 0; l < m; l++)
						if(data[k][l] && (k != i || l != j))
							ans++;
	return ans;
}

int abs_int(int x){
	return x < 0 ? -x : x;
}

long long count2(){
	long long ans = 0LL;
	int i, j, k, l, level, dist;
	for(i = 0; i < n; i++)
		for(j = 0; j < m; j++)
			if(data[i][j]){
				level = inf;
				for(k = 0; k < n; k++)
					for(l = 0; l < m; l++)
						if(!data[k][l]){
							dist = abs_int(k - i) + abs_int(l - j);
							if(level > dist)
								level = dist;
						}
				ans += level;
			}
	return ans;
}

int main(){
	int i;
	scanf("%d", &seed);
	for(i = 0; i < 10; i++){
		generate_input();
		printf("%lld\n", type == 1 ? count2() : count1());
	}
	return 0;
}
